codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/endpoint/EndpointParamsGenerator.kt

```diff
@@ -0,0 +1,278 @@
+import software.amazon.smithy.rust.codegen.core.rustlang.isCopy
+import software.amazon.smithy.rust.codegen.core.rustlang.rust
+import software.amazon.smithy.rust.codegen.core.rustlang.rustBlock
+import software.amazon.smithy.rust.codegen.core.rustlang.rustBlockTemplate
+import software.amazon.smithy.rust.codegen.core.rustlang.rustTemplate
+import software.amazon.smithy.rust.codegen.core.rustlang.stripOuter
+import software.amazon.smithy.rust.codegen.core.rustlang.writable
+import software.amazon.smithy.rust.codegen.core.smithy.RuntimeType
+import software.amazon.smithy.rust.codegen.core.smithy.RuntimeType.Companion.Clone
+import software.amazon.smithy.rust.codegen.core.smithy.RuntimeType.Companion.Debug
+import software.amazon.smithy.rust.codegen.core.smithy.RuntimeType.Companion.Default
+import software.amazon.smithy.rust.codegen.core.smithy.RuntimeType.Companion.PartialEq
+import software.amazon.smithy.rust.codegen.core.smithy.isOptional
+import software.amazon.smithy.rust.codegen.core.smithy.makeOptional
+import software.amazon.smithy.rust.codegen.core.smithy.mapRustType
+import software.amazon.smithy.rust.codegen.core.smithy.rustType
+import software.amazon.smithy.rust.codegen.core.util.dq
+import software.amazon.smithy.rust.codegen.core.util.orNull
+
+// TODO(https://github.com/awslabs/smithy-rs/issues/1927): When endpoint resolution is implemented, remove doc-hidden
+val EndpointsModule = RustModule.public("endpoint", "Endpoint resolution functionality")
+    .copy(rustMetadata = RustMetadata(additionalAttributes = listOf(Attribute.DocHidden), visibility = Visibility.PUBLIC))
+
+/** Endpoint Parameters generator.
+ *
+ * This class generates the `Params` struct for an [EndpointRuleset]. The struct will have `pub(crate)` fields, a `Builder`,
+ * and an error type that is created to handle when construction fails.
+ *
+ * The builder of this struct generates a fallible `build()` method because endpoint params MAY have required fields. However,
+ * the external parts of this struct (the public accessors) will _always_ be optional to ensure a public interface is maintained.
+ *
+ * The following snippet contains an example of what is generated (eliding the error):
+ *  ```rust
+ *  #[non_exhaustive]
+ *  #[derive(std::clone::Clone, std::cmp::PartialEq, std::fmt::Debug)]
+ *  /// Configuration parameters for resolving the correct endpoint
+ *  pub struct Params {
+ *      pub(crate) region: std::option::Option<std::string::String>,
+ *  }
+ *  impl Params {
+ *      /// Create a builder for [`Params`]
+ *      pub fn builder() -> crate::endpoint_resolver::Builder {
+ *          crate::endpoint_resolver::Builder::default()
+ *      }
+ *      /// Gets the value for region
+ *      pub fn region(&self) -> std::option::Option<&str> {
+ *          self.region.as_deref()
+ *      }
+ *  }
+ *
+ *  /// Builder for [`Params`]
+ *  #[derive(std::default::Default, std::clone::Clone, std::cmp::PartialEq, std::fmt::Debug)]
+ *  pub struct Builder {
```

Comment:
Will there ever be any other kind of builder in this module? Should it be `ParamsBuilder`?

Reply:
probably not, but that's a clearer name